lib/srv/server/installer/autodiscover.go

```diff
@@ -0,0 +1,600 @@
+				clt, err := awsimds.NewInstanceMetadataClient(ctx)
+				return clt, trace.Wrap(err)
+			},
+			func(ctx context.Context) (imds.Client, error) {
+				return azureimds.NewInstanceMetadataClient(), nil
+			},
+			func(ctx context.Context) (imds.Client, error) {
+				instancesClient, err := gcp.NewInstancesClient(ctx)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+
+				clt, err := gcpimds.NewInstanceMetadataClient(instancesClient)
+				return clt, trace.Wrap(err)
+			},
+		}
+	}
+
+	return nil
+}
+
+type packageVersion struct {
+	Name    string
+	Version string
+}
+
+type packageManager interface {
+	AddTeleportRepository(ctx context.Context, ldi *linuxDistroInfo, repoChannel string) error
+	InstallPackages(context.Context, []packageVersion) error
+}
+
+// AutodiscoverNodeInstaller will install teleport in the current system.
+// It's meant to be used by the Server Auto Discover script.
+type AutodiscoverNodeInstaller struct {
+	*AutodiscoverNodeInstallerConfig
+}
+
+// NewAutodiscoverNodeInstaller returns a new AutodiscoverNodeInstaller.
+func NewAutodiscoverNodeInstaller(cfg *AutodiscoverNodeInstallerConfig) (*AutodiscoverNodeInstaller, error) {
+	if err := cfg.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ti := &AutodiscoverNodeInstaller{
+		AutodiscoverNodeInstallerConfig: cfg,
+	}
+
+	return ti, nil
+}
+
+const (
+	// exclusiveInstallFileLock is the name of the lockfile to be used when installing teleport.
+	// Used for the default installers (see api/types/installers/{agentless,}installer.sh.tmpl/).
+	exclusiveInstallFileLock = "/var/lock/teleport_install.lock"
+
+	// etcOSReleaseFile is the location of the OS Release information.
+	// This is valid for most linux distros, that rely on systemd.
+	etcOSReleaseFile = "/etc/os-release"
+)
+
+var imdsClientTypeToJoinMethod = map[types.InstanceMetadataType]types.JoinMethod{
+	types.InstanceMetadataTypeAzure: types.JoinMethodAzure,
+	types.InstanceMetadataTypeEC2:   types.JoinMethodIAM,
+	types.InstanceMetadataTypeGCP:   types.JoinMethodGCP,
+}
+
+// Install teleport in the current system.
+func (asi *AutodiscoverNodeInstaller) Install(ctx context.Context) error {
+	// Ensure only one installer is running by locking the same file as the script installers.
+	lockFile := asi.buildAbsoluteFilePath(exclusiveInstallFileLock)
+	unlockFn, err := utils.FSTryWriteLock(lockFile)
+	if err != nil {
+		return trace.BadParameter("Could not get lock %s. Either remove it or wait for the other installer to finish.", lockFile)
+	}
+	defer func() {
+		if err := unlockFn(); err != nil {
+			asi.Logger.WarnContext(ctx, "Failed to remove lock. Please remove it manually.", "file", exclusiveInstallFileLock)
+		}
+	}()
+
+	// Check if teleport is already installed.
+	if _, err := os.Stat(asi.binariesLocation.teleport); err == nil {
+		asi.Logger.InfoContext(ctx, "Teleport is already installed in the system.")
+		return nil
+	}
```

Comment:
Can it happen that teleport binary is present but we failed to generate/write config (or it got removed for some reason)? In this case the method will not proceed but installation will not be valid.

Or are you planning to address this scenario in a follow-up PR?

Reply:
Yes, I'm planning on multiple follow-up PRs
I'm tracking improvements on the 1st tasklist of this issue https://github.com/gravitational/teleport/issues/37620

Feel free to add more improvements there 👍